### DISCUSSION TITLE
ModelSerializer inconsistently serializes UUID foreign keys

### DISCUSSION THREAD

**Discussion by r-thomson** [encode/django-rest-framework#8786] [Potential Issue]
I'm working in a REST Framework app that's using UUIDs for primary keys, and I've run into some unexpected behavior with the way `ModelSerializer` handles foreign keys. I think it's worth considering changing this behavior to make things more consistent.

#### Model File

```py
from django.db import models

class PrimaryModel(models.Model):
    id = models.UUIDField(primary_key=True)
    related = models.ForeignKey('RelatedModel', on_delete=models.DO_NOTHING)

class RelatedModel(models.Model):
    id = models.UUIDField(primary_key=True)
```

#### Serializer File

```py
from rest_framework import serializers

class ExampleSerializer(serializers.ModelSerializer):
    class Meta:
        model = PrimaryModel
        fields = ('id', 'related', 'related_id')
```

If we look at the fields that get generated by `ModelSerializer`, we can see that the fields are mapped as so:

* `id` (`models.UUIDField`) becomes `serializers.UUIDField`
* `related` becomes `serializers.PrimaryKeyRelatedField`
* `related_id` becomes `serializers.ReadOnlyField`

```py
ExampleSerializer():
    id = UUIDField(validators=[...])
    related = PrimaryKeyRelatedField(queryset=RelatedModel.objects.all())
    related_model = ReadOnlyField()
```

If we serialize a `PrimaryModel` instance, we can see that we actually get different types in the response: `id` becomes a `str`, while the rest become `UUID`s.

```py
{
    'id': 'a04f3708-35b7-4ae6-98c5-56e7a598de85',
    'related': UUID('1e5aa453-b351-4d56-a2c1-cf5697a4892b'),
    'related_id': UUID('6100547b-f0ac-4ab2-be8c-23076fee2741')
}
```

In practice, `JSONRenderer` will convert any `UUID` object to strings, so it makes no difference for the public API. However, it definitely introduces some friction when working with the serializer data directly, like when writing tests.

## Suggested Solution

I think the most sensible behavior here would be for the `pk_field` kwarg on `PrimaryKeyRelatedField` to be automatically set to `serializers.UUIDField` if the foreign key's type is a UUID, unless the user explicitly sets it to something else. This could be controlled by a setting to avoid breaking changes.

While the behavior of `related_id` here is still a bit of a gotcha, I think it ultimately makes sense given the behavior of `ReadOnlyField`.

**Comment by auvipy** [comment]
That looks interesting! can you start a PR with a failing test case?

**Reply by r-thomson** [reply]
See https://github.com/encode/django-rest-framework/pull/8791

I can also take a crack at implementation, if you’d like

**Reply by auvipy** [reply]
you love to see that as well. considering that it will fix the bug and not introduce any new API surface and won't create major regression

**Reply by r-thomson** [reply]
https://github.com/encode/django-rest-framework/pull/8791 updated 👍
